From eccc3536e12a86434ced1529fcad5bbbb280ffeb Mon Sep 17 00:00:00 2001
From: Ryan Walklin <ryan@testtoast.com>
Date: Sat, 21 Dec 2024 23:01:14 +1300
Subject: [PATCH] sun4i-codec: add routing for headphones and internal speaker

The sun4i-codec kernel driver covers a large number of Allwinner SoCs.
The H616 codec has a single line-out route, which is used in concert
with a Toshiba mux chip to send audio to either an internal speaker or
headphone jack on a number of Anbernic handheld gaming devices.

Add a UCM configuration to allow enabling/disabling the speaker amp
depending on whether headphones are in use or not.

Signed-off-by: Ryan Walklin <ryan@testtoast.com>
---
 ucm2/Allwinner/sun4i-h616/HiFi.conf           | 35 +++++++++++++++++++
 ucm2/conf.d/sun4i-codec/h616-audio-codec.conf |  8 +++++
 2 files changed, 43 insertions(+)
 create mode 100644 ucm2/Allwinner/sun4i-h616/HiFi.conf
 create mode 100644 ucm2/conf.d/sun4i-codec/h616-audio-codec.conf

diff --git a/ucm2/Allwinner/sun4i-h616/HiFi.conf b/ucm2/Allwinner/sun4i-h616/HiFi.conf
new file mode 100644
index 00000000..2cb9a31f
--- /dev/null
+++ b/ucm2/Allwinner/sun4i-h616/HiFi.conf
@@ -0,0 +1,40 @@
+SectionDevice."Speaker" {
+	Comment "Internal Speaker"
+
+	EnableSequence [
+		cset "name='DAC Playback Volume' 63"
+		cset "name='Speaker Switch' on"
+	]
+
+	Value {
+		PlaybackMixerElem "Line Out"
+		PlaybackPCM "hw:${CardId},0"
+	}
+
+	DisableSequence [
+		cset "name='Speaker Switch' off"
+	]
+
+	ConflictingDevice [
+		"Headphones"
+	]
+}
+
+
+SectionDevice."Headphones" {
+	Comment "Headphones"
+
+	EnableSequence [
+		cset "name='DAC Playback Volume' 63"
+	]
+
+	Value {
+		PlaybackMixerElem "Line Out"
+		PlaybackPCM "hw:${CardId},0"
+		JackControl "Headphone Jack"
+	}
+
+	ConflictingDevice [
+		"Speaker"
+	]
+}
diff --git a/ucm2/conf.d/sun4i-codec/h616-audio-codec.conf b/ucm2/conf.d/sun4i-codec/h616-audio-codec.conf
new file mode 100644
index 00000000..240f41ab
--- /dev/null
+++ b/ucm2/conf.d/sun4i-codec/h616-audio-codec.conf
@@ -0,0 +1,12 @@
+# UCM for Allwinner H616 Codec
+
+Syntax 3
+
+SectionUseCase."HiFi" {
+	File "/Allwinner/sun4i-h616/HiFi.conf"
+	Comment "Play HiFi quality Music"
+}
+
+FixedBootSequence [
+	cset "name='DAC Playback Switch' on"
+]
